### ISSUE TITLE
crypto/sha256: Implement optimized SSE based in plan9

### ISSUE THREAD

**Issue by vadmeste** [minio/minio#1576]
SHA extensions available starting from Skylake: https://en.wikipedia.org/wiki/Intel_SHA_extensions

**Comment by harshavardhana** [comment]
https://software.intel.com/en-us/articles/intel-sha-extensions - relevant article.

**Comment by GarimaKapoor** [comment]
Frank Wessels @fwessels is looking into this.

**Comment by fwessels** [comment]
As of yet there are unfortunately no CPUs that support these extensions (supposedly coming in 2H of this year or 2017).

Linux 4.4 apparently added support that we could look at how they implemented it (but no idea how they tested this): 
http://lkml.iu.edu/hypermail/linux/kernel/1511.0/00383.html

And now with the SIMD optimized BLAKE2 this is also less of an issue.

So I suggest postponing the issue and revisit it in due course to see if it gives substantial performance benefits.

**Comment by harshavardhana** [comment]
> And now with the SIMD optimized BLAKE2 this is also less of an issue.
> 
> So I suggest postponing the issue and revisit it in due course to see if it gives substantial performance benefits.

The reason i added this for distributed is because sha256 being comonly used for signature verification - https://golang.org/src/crypto/sha256/sha256block_amd64.s

Would be great if we can convert sha256 to also have SSE capabilities. https://github.com/wereHamster/sha256-sse/blob/master/sha256-sse.c 

We do have GAS code though https://github.com/minio/minio/tree/master/pkg/crypto/sha256 which was previously done to use CGO based, but we saw issues when CGO based sha256 was used under load perhaps there are some known issues in this, so we moved default `cryptos/sha256` for now. 

We should evaluate and see if it is possible..

**Comment by harshavardhana** [comment]
Moved the context of the bug to be not tied to specific instruction sets.

**Comment by fwessels** [comment]
Ok, great.
Will look into this, get a sense for the speed up we can achieve for the different architectures.

**Comment by harshavardhana** [comment]
```
$ benchcmp /tmp/golang.txt /tmp/assembly.txt 
benchmark                 old ns/op     new ns/op     delta
BenchmarkHash8Bytes-4     406           1508          +271.43%
BenchmarkHash1K-4         5956          5683          -4.58%
BenchmarkHash8K-4         41351         30565         -26.08%
BenchmarkHash1M-4         5726447       3767999       -34.20%

benchmark                 old MB/s     new MB/s     speedup
BenchmarkHash8Bytes-4     19.67        5.30         0.27x
BenchmarkHash1K-4         171.91       180.18       1.05x
BenchmarkHash8K-4         198.11       268.02       1.35x
BenchmarkHash1M-4         183.11       278.28       1.52x
$ 
```

Current CGO based v/s Golang standard package. CGO is miserable for small blocks.

**Comment by fwessels** [comment]
Nice benchmark and poor performance for small blocks indeed. 

Do you know which version was used: SSE or AVX/2?

**Comment by harshavardhana** [comment]
i think it used AVX based on my laptop @fwessels

**Comment by fwessels** [comment]
Looked at https://github.com/wereHamster/sha256-sse/blob/master/sha256-sse.c which gives:

```
$ ./test
3193.62 blocks per second
8526.52 blocks per second
```

First result is reference implementation (C), second result is SSE, but overall around 64M/sec which isn't great compared to eg. openssl:

```
$ openssl speed sha256
Doing sha256 for 3s on 16 size blocks: 6265003 sha256's in 3.00s
Doing sha256 for 3s on 64 size blocks: 3599770 sha256's in 2.99s
Doing sha256 for 3s on 256 size blocks: 1613882 sha256's in 3.00s
Doing sha256 for 3s on 1024 size blocks: 512219 sha256's in 3.00s
Doing sha256 for 3s on 8192 size blocks: 68405 sha256's in 3.00s
OpenSSL 0.9.8zg 14 July 2015
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes
sha256           33413.02k    77038.64k   137691.36k   174966.43k   186844.94k
```

180 MB/sec is comparable to what we get in Golang right now.
Investigating further...

**Comment by fwessels** [comment]
Did some more benchmarks with https://github.com/minio/minio/tree/master/pkg/crypto/sha256 on minio3, first four columns are in MB/sec, last 3 columns compare to standard golang implementation (which is in assembler but not SIMD optimized):

```
                        golang (std)   SSSE3     AVX    AVX2     SSSE3    AVX   AVX2
BenchmarkHash8Bytes-12          17.6     6.5     6.5     6.1      0.37   0.37   0.35
BenchmarkHash1K-12             170.1   159.5   159.4   173.6      0.94   0.94   1.02
BenchmarkHash8K-12             187.3   265.9   275.8   300.1      1.42   1.47   1.60
BenchmarkHash1M-12             189.2   286.6   306.1   368.0      1.51   1.62   1.95
```

Clearly the SIMD versions do poorly for very small sizes, and achieve up to a 2x improvement (AVX2) for larger sizes.

As per #1995 considerable time is spent in SHA256 calculations, so it makes sense to migrate to Plan 9 assembly, unless almost all time is spent on small sizes only.

Even if SIMD optimized versions continue to do poorly for small objects, we can put in a heuristic that for sizes < 512 (or whatever we choose) we default back to the golang implementation.

**Comment by harshavardhana** [comment]
> As per #1995 considerable time is spent in SHA256 calculations, so it makes sense to migrate to Plan 9 assembly, unless almost all time is spent on small sizes only.
> 
> Even if SIMD optimized versions continue to do poorly for small objects, we can put in a heuristic that for sizes < 512 (or whatever we choose) we default back to the golang implementation.

Another thing to note is that CGO implementation is kind of chattier of the CPU so could cause its own lag, we never really went deeper to figure out why or may be the assembly we are using is not written/optimized for it. 

BTW the assembly code is borrowed from Linux kernel crypto implementation.  Plan9 based would be perhaps faster than the numbers here.

**Comment by fwessels** [comment]
Initial support for AVX added with following results (see minio/sha256-simd):

```
$ benchcmp golang.txt avx.txt 
benchmark                 old ns/op     new ns/op     delta
BenchmarkHash8Bytes-8     429           327           -23.78%
BenchmarkHash1K-8         5608          3823          -31.83%
BenchmarkHash8K-8         41813         27980         -33.08%
BenchmarkHash1M-8         5272793       3567815       -32.34%

benchmark                 old MB/s     new MB/s     speedup
BenchmarkHash8Bytes-8     18.61        24.44        1.31x
BenchmarkHash1K-8         182.56       267.78       1.47x
BenchmarkHash8K-8         195.92       292.77       1.49x
BenchmarkHash1M-8         198.87       293.90       1.48x
```

Luckily we are not seeing the drop in speed for small sizes (as compared to the cgo based version)

**Comment by fwessels** [comment]
Benchmarks for SSSE (on MacBook Pro)

```
$ benchcmp golang.txt ssse.txt 
benchmark                 old ns/op     new ns/op     delta
BenchmarkHash8Bytes-8     429           348           -18.88%
BenchmarkHash1K-8         5608          4015          -28.41%
BenchmarkHash8K-8         41813         29738         -28.88%
BenchmarkHash1M-8         5272793       3775766       -28.39%

benchmark                 old MB/s     new MB/s     speedup
BenchmarkHash8Bytes-8     18.61        22.93        1.23x
BenchmarkHash1K-8         182.56       255.00       1.40x
BenchmarkHash8K-8         195.92       275.47       1.41x
BenchmarkHash1M-8         198.87       277.71       1.40x
```

**Comment by fwessels** [comment]
SHA256 performance experiment from bitcoin:
https://github.com/laanwj/bitcoin/tree/2016_05_sha256_accel

Giving more or less consistent results with what we are getting

**Comment by fwessels** [comment]
Support for AVX2 had been added in minio/sha256-simd.
Would be good to give this a good stress test in Minio server to see how it performs and holds out.

**Comment by harshavardhana** [comment]
> Support for AVX2 had been added in minio/sha256-simd.
> Would be good to give this a good stress test in Minio server to see how it performs and holds ou

You can vendorize with 'make pkg-add PKG=github.com/minio/sha256-simd'  for the server , we will start testing the benefits right away. :-)

**Comment by harshavardhana** [comment]
This is fixed #2320
